classes: Dedupe all classes

TimothyGu opened the pull request:
This causes a slow down especially on cases with simple string-only arrays (~3×), but this is the best I can do. Optimization welcome.

TimothyGu committed:
classes: Dedupe all classes

This causes a slow down especially on cases with simple string-only
arrays, but this is the best I can do. Optimization welcome.

alubbe commented:
For escapeAlways, couldn't we just check equality to true?

In terms of performance, I ran some benchmarks and the only to speed it up
would to be to use ES6 Sets instead of objects. We could feature detect
them, node has them, more and more browsers have them and we could fall
back to objects.
Am 13.01.2016 4:30 vorm. schrieb "Timothy Gu" notifications@github.com:

> This causes a slow down especially on cases with simple string-only arrays
> 
> ## (~3×), but this is the best I can do. Optimization welcome.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/pugjs/pug-runtime/pull/25
> Commit Summary
> - classes: Dedupe all classes
> 
> File Changes
> - _M_ index.js
>   https://github.com/pugjs/pug-runtime/pull/25/files#diff-0 (53)
> 
> Patch Links:
> - https://github.com/pugjs/pug-runtime/pull/25.patch
> - https://github.com/pugjs/pug-runtime/pull/25.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pugjs/pug-runtime/pull/25.

ForbesLindesay commented:
One option would be to enable only in dev-mode and throw errors rather than deduping?

TimothyGu commented:
> One option would be to enable only in dev-mode and throw errors rather than deduping?

Err… I don't see the point in that. Throwing errors prevents a perfectly okay use-case that allows more flexibility. Plus, to throw errors we have to first check for dups, so no performance gains can be reaped from this approach either.

TimothyGu commented:
> In terms of performance, I ran some benchmarks and the only to speed it up would to be to use ES6 Sets instead of objects.

I did benchmark Sets but it was done on an older, slower version of the patch that didn't share the output object, so no performance gains were achieved. I'll test some more.

TimothyGu commented:
Nah, Set is still a lot slower. I also added another optimization that prevents a branch `out = out || {}`.

> For escapeAlways, couldn't we just check equality to true?

Derp. Fixed.

ForbesLindesay commented:
The advantage of throwing on dupes is that we could do all the deduping inside `if (process.env.NODE_ENV !== 'production')`, which would mean no performance penalty in production.

I'm not saying it's the best thing to do, just that it's an option.

alubbe commented:
That's a good point - do we actually want to check for dupes? It is valid html to print classes more than once and we do take a hit for these extra checks.
